tianshou/data/batch.py

```diff
@@ -177,12 +176,21 @@ def __init__(self,
             elif _is_batch_set(batch_dict):
                 self.stack_(batch_dict)
         if len(kwargs) > 0:
-            self.__init__(kwargs, copy=copy)
+            self.__init__(kwargs, copy=copy)  # type: ignore
 
     def __setattr__(self, key: str, value: Any) -> None:
         """self.key = value"""
         self.__dict__[key] = _parse_value(value)
 
+    def __getattr__(self, key: str) -> Any:
```

Comment:
Override `__getattr__` can directly inherit existing functions of dict, thus we do not need to rewrite keys/values/items/...

Reply:
I'm strongly against overloading standard methods not to change anything but for type checking. It is extremely confusing to me. It should be possible to better configure the type checker, otherwise better dropping it since it is probably not good. Those methods in particular are very critical and I feel very uncomfortable to overload them for nothing.

Reply:
No, it isn't only for type checking. If we use something like `batch.pop` it uses the pop function in `dict`.
Also, during my interaction with mypy, it cannot configure to remove this attribute error. Only a `getattr -> Any` can make it shut up. Otherwise, we need to add `type: ignore` everywhere. See https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html

```python
# If you want dynamic attributes on your class, have it override "__setattr__"
# or "__getattr__" in a stub or in your source code.
#
# "__setattr__" allows for dynamic assignment to names
# "__getattr__" allows for dynamic access to names
class A:
    # This will allow assignment to any A.x, if x is the same type as "value"
    # (use "value: Any" to allow arbitrary types)
    def __setattr__(self, name: str, value: int) -> None: ...

    # This will allow access to any A.x, if x is compatible with the return type
    def __getattr__(self, name: str) -> int: ...

a.foo = 42  # Works
a.bar = 'Ex-parrot'  # Fails type checking
```

This is the example described in their cheatsheet.

Reply:
I tested and in fact, whether to overload getattr does not hurt the performance.

Reply:
It is not about performance but clarity. Overloading a method for nothing is just very confusing. Type checking is supposed to make things more clear, not the contrary.

Reply:
But if you are overloading `__getattr__` on purpose, for instance to benefit from some standard methods, I don't think it is very good either. Ideally it would be better to find a way to inherit from dict directly in a way. But for now I guess it is good enough.